### DISCUSSION TITLE
Can I convert from pix to PIL without saving to disk?

### DISCUSSION THREAD

**Discussion by alejandrofm** [pymupdf/PyMuPDF#1678] [Q&A] [answered]
Hi! I am currently doing this to extract images from PDFs and then feeding those to tesseract-OCR, it seems that I'm wasting, time, quality and not optimizing the code when I write to PNG and read the PNG in the next line, is there any way to feed a pixmap to a Image element?

                pix = page.getPixmap(matrix=mat, alpha=False)  # render page to an image
                saving_name = filepath.split('\\')[-1].replace('pdf', '') # extract name from file
                pix.writePNG("page{}{}.png".format(saving_name, page.number))  # store image as a PNG
                imagen = Image.open("page{}{}.png".format(saving_name, page.number))  # open image
                imagen.show() #show image

Thank you!

**Comment by JorjMcKie** [comment]
Sure there is!
There are even a number of alternatives to avoid it:

1. You can convert the pixmap to a number of image formats as bytes objects. Use ``data = pix.getImageData("format")`` for this. You can choose among several "format" alternatives trading size of the bytes object ``data`` for creation speed. Then use ``img = Image.open(io.BytesIO(data), ...)`` to open it as a PIL image.

2. Use the pixmap directly as input to create a PIL image: ``img = Image.frombytes(mode, [pix.width, pix.height], pix.samples)``. The string "mode" in your case should probably be "RBG".

To improve the OCR detection, you can render the PDF page at higher quality by using a matrix in ``Page.getPixmap()``. E.g. ``matrix = fitz.Matrix(2,2)`` will increase resolution by a factor of 2 in each (x, y) direction - thus yileding a 4 times larger pixmap.
But if the PDF contains images (which I assume to be) then that may have limited effect for OCR recognition. If on the other hand the page contains normal text, then there would be no need to use OCR ...

**Reply by Aspace2create** [reply]
line 146, in keyPressEvent
    data=pxm.getImageData("PNG")
AttributeError: 'QPixmap' object has no attribute 'getImageData'

**Comment by alejandrofm** [comment]
Thank you @JorjMcKie its a single image per page and I'm already using matrix 2,2 with great results.
Any option (1 or 2) or parameter recommendation for the conversion to PIL?
Thanks again!

**Comment by JorjMcKie** [comment]
Option 2 is about one third faster than option 1 with a similar memory footprint ...
I just measured ``10 loops, best of 3: 42.3 ms per loop`` (option 2) versus ``10 loops, best of 3: 64.5 ms per loop`` (option 1) for a large pixmap with a samples size of about 23 MB.

Option 1 is important for feeding image data to some packages, most notably ``tkinter``. In the latter case you would use ``pix.getImageData("ppm")`` as input to ``tk.PhotoImage`` in a Python version independent manner.

**Comment by alejandrofm** [comment]
Thankyou A LOT @JorjMcKie it works great, could I have any problem if the source image is grayscale or b&w? will it "scale" to RGB?
Thanks!

**Comment by JorjMcKie** [comment]
You are welcome!
No, you can freely choose the resulting colorspace when using ``getPixmap``. If using ``fitz.csGRAY`` the pixmap will be smaller of course (only one third of RGB), which may be utterly sufficient for doing the OCR stuff.

**Comment by alejandrofm** [comment]
I'm VERY happy with the result currently using `fitz.csGRAY` and mode `R` , with your help I improved:

- Avoid creating a single a file for each page and writing it to disk.
- Avoid reading file from disk in next line of code.
- Lower size of the file i'm working on.

Thank you a lot!

**Comment by JorjMcKie** [comment]
close issue?

**Comment by rozeappletree** [comment, marked as answer]
@alejandrofm 

> I'm VERY happy with the result currently using `fitz.csGRAY` and mode `R` , with your help I improved:
> 
> * Avoid creating a single a file for each page and writing it to disk.
> * Avoid reading file from disk in next line of code.
> * Lower size of the file i'm working on.
> 
> Thank you a lot!

Can you please give a code snippet for `fitz.csGRAY` and `mode R`

**Comment by JorjMcKie** [comment]
@rakesh4real @alejandrofm -
Just wanted to let you guys know, that the recent PyMuPDF versions have new methods for direct support of PIL / Pillow - so you no longer need fiddling around with pixmap attributes to achieve this:
* ``Pixmap.pillowSave(...)`` saves a pixmap to an image file using Pillow. The arguments are the same as for ``Image.save(...)`` in pillow. This makes it possible, to output JPEG images for example ... just everything supported by Pillow.
Note that this method will be renamed to ``pixmap.pil_save()`` in version 1.18.14.
* Similarly, ``Pixmap.pillowData(...)`` returns a ``bytes`` object using Pillow. Otherwise the same arguments supported.
Again a rename in version 1.18.14: ``Pixmap.pil_tobytes()``.

**Comment by canklot** [comment]
For those who wants numpy array not PIL.Image

```
pix = PIL.Image.frombytes("RGB", [pix.width, pix.height], pix.samples)
pix = np.array(pix)
```

**Comment by JorjMcKie** [comment]
@canklot - there is an even more direct and much faster way for ndarrays: see [here](https://github.com/pymupdf/PyMuPDF/discussions/1208).
